src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.BoxedTypes.cs

```diff
@@ -415,6 +428,65 @@ public override MethodIL EmitIL()
             }
         }
 
+        /// <summary>
+        /// Represents a thunk to call instance method on boxed valuetypes.
+        /// </summary>
+        private partial class UnboxingThunk : ILStubMethod
+        {
+            private MethodDesc _targetMethod;
+            private BoxedValueType _owningType;
+
+            public UnboxingThunk(BoxedValueType owningType, MethodDesc targetMethod)
+            {
+                Debug.Assert(targetMethod.OwningType.IsValueType);
+                Debug.Assert(!targetMethod.Signature.IsStatic);
+
+                _owningType = owningType;
+                _targetMethod = targetMethod;
+            }
+
+            public override TypeSystemContext Context => _targetMethod.Context;
+
+            public override TypeDesc OwningType => _owningType;
+
+            public override MethodSignature Signature => _targetMethod.Signature;
+
+            public MethodDesc TargetMethod => _targetMethod;
+
+            public override string Name
+            {
+                get
+                {
+                    return _targetMethod.Name + "_Unbox";
+                }
+            }
+
+            public override MethodIL EmitIL()
+            {
+                // Generate the unboxing stub. This loosely corresponds to following C#:
+                // return BoxedValue.InstanceMethod([rest of parameters])
+
+                ILEmitter emit = new ILEmitter();
+                ILCodeStream codeStream = emit.NewCodeStream();
+
+                // unbox to get a pointer to the value type
+                codeStream.EmitLdArg(0);
+                codeStream.Emit(ILOpcode.unbox, emit.NewToken(_owningType));
```

Comment:
This will throw an `InvalidCastException`. Owning type is not the valuetype, it's a "type with the same layout as a boxed valuetype". I would prefer we use the `ldflda` approach the special unbox thunk uses because it avoids the typecheck completely and codegens are much better at optimizing it.